Implement user sessions

ajbozarth opened the pull request:
### Overview

This is a implementation of user sessions. A user session is determined by a cookie set in thier browser after cookie consent has been accepted (via cookie consent banner) and is unique to each instance of the web app. 

User's can only see, interact with and delete images uploaded with their cookie. The user id cookie is signed and encrypted server-side for security, but anyone with server access can see all user images while they exist on the server.

I also added a delete uploaded images button to the main page since it now only affects the current user. Given this, it made sense to remove the cleanup page from the app since it's functionality had moved to the main page.

The tests have also been updated to work with cookies and test multiple user use cases.

File clean up has also been added with user uploaded files deleted after one day. The expired files clean up is run on any main page load (not on other endpoints).

This implementation should meet current EU cookie consent requirements.

### Included Bug Fixes

- No longer fails to upload files with a `.` in their filename.

- Detail pages work for images with a space in their filename now.

### Purpose

The overall purpose of this update is allow for long running instances that support multiple users.

ajbozarth commented:
I figured since its EOD that I should put this up for review while I finish working on user file cleanup (since they probably won't clean up after themselves).

ajbozarth commented:
Also @ptitzler and @bdwyer2 if you want to take a look at this for an extra set of eye on the python code

MLnick commented on app.py:
```diff
@@ -47,35 +48,45 @@
+    def prepare(self):
+        if not self.get_cookie(app_cookie):
+            user_id = str(uuid.uuid4())
+            self.set_cookie(app_cookie, str(uuid.uuid4()))
```
`set_cookie` should set `user_id` right? Here it's generating a new UUID

ajbozarth replied:
Nice catch, the error here was actually that I was logging user_id then creating a new one `str(uuid.uuid4())` still instead of using it. In this implementation user_id is simply the value of the `max-image-caption-generator-web-app` cookie in their web-browser

ajbozarth commented:
I just pushed the file clean up code. That means this PR is no longer WIP and ready to merge once it passes review

ajbozarth commented:
/@all: I'd appreciate any feedback here, I want to make sure my python is good and that the UX is solid. If you don't have time to deploy the change to test it it's also running at http://max-demo1.fyre.ibm.com:8088 I'd like to merge this early this week so I can update the long running cluster in time for this week's release.

ajbozarth commented on app.py:
```diff
@@ -119,13 +133,33 @@ def post(self):
         self.finish(json.dumps(finish_ret))
 
 
+def get_user_img_prefix(self):
```
No, it's a helper function, why would it be in a Handler class?

djalova replied:
Seems like it's just being used in the Handler and its subclasses.

ajbozarth replied:
Oh I get where you're coming from, the Handler classes are only meant to have the override functions (get, post, etc) in them when using tornado, any other functions should be global, at least that's my understanding based on example tornado code

stevemart commented on app.py:
```diff
@@ -39,43 +40,55 @@
 logging.basicConfig(level=os.environ.get("LOGLEVEL", "INFO"),
-                    format='%(levelname)s: %(message)s')
+                    format='[%(asctime)s] %(levelname)s: %(message)s',
+                    datefmt='%m/%d/%y %H:%M:%S')
```
you don't have to actually submit a `datefmt` parameter just cause you're using `asctime`. it defaults to ISO-8601 standard (https://docs.python.org/2/howto/logging.html#displaying-the-date-time-in-messages) which is `datefmt='%m/%d/%Y %I:%M:%S %p'`, and just slightly different from what you're doing. Of course, if you have a particular reason for doing it your way, that's cool, too.

ajbozarth replied:
Short answer is I like military time in my logs, long answer is I played around with the format quite a bit and didn't realized I settled on something so close to the default in the end.

stevemart commented on app.py:
```diff
@@ -39,43 +40,55 @@
-class MainHandler(web.RequestHandler):
+class BaseHandler(web.RequestHandler):
+    def prepare(self):
+        if not self.get_cookie(app_cookie):
```
might be worth adding a link to http://www.tornadoweb.org/en/branch2.4/web.html#cookies as a comment here

ajbozarth replied:
Will do

ajbozarth replied:
Using the Doc link instead: http://www.tornadoweb.org/en/stable/guide/security.html

stevemart commented on app.py:
```diff
@@ -189,7 +238,7 @@ def signal_handler(sig, frame):
 def shutdown():
     logging.info("Cleaning up image files")
-    clean_up()
+    clean_up_user_images(None)
```
why pass in `None` here?

stevemart replied:
also, did you want to clear the cookie here?

ajbozarth replied:
If `self` isn't set (ie `= None`) then it clears ALL temp file, not just the requesting user, which is what I want on shutdown. As for why `None` and not empty, python gives an error if I didn't pass `None`. And there's no cookie to clear here, it's a server side function, this is called when a user `Ctrl-C`s the app

stevemart commented on app.py:
```diff
@@ -174,13 +208,28 @@ def prepare_metadata():
+    exp_time = time.time() - (24 * 60 * 60)
     for img_file in img_list:
-        if img_file.startswith(static_img_path + temp_img_prefix):
+        if img_file.startswith(static_img_path + temp_img_prefix)\
```
nit, i don't like using slashes, you can use parens:

```
        if (img_file.startswith(static_img_path + temp_img_prefix)
                and os.stat(img_file).st_ctime < exp_time):
```

ajbozarth replied:
thank you, I couldn't remember how to get it to work with out the `slash` that my IDE kept adding.

stevemart commented on app.py:
```diff
@@ -174,13 +208,28 @@ def prepare_metadata():
 
-# Deletes all files uploaded through the GUI and removes them from the dict
-def clean_up():
+# Deletes files uploaded through the GUI and removes them from the dict
```
Why not use actual docstring format? https://www.python.org/dev/peps/pep-0257/

```
def clean_up_user_images(self):
    """Cleans user images.

    Deletes files uploaded through the GUI and removes them from the dict
    If 'self' is given then only the current user's images are deleted
    """
    img_prefix = get_user_img_prefix(self) if self else temp_img_prefix
```

ajbozarth replied:
No experience with doc string, I'll switch to it, I assume it's a better standard

stevemart commented on app.py:
```diff
@@ -174,13 +208,28 @@ def prepare_metadata():
-def clean_up():
+# Deletes files uploaded through the GUI and removes them from the dict
+# If 'self' is given then only the current user's images are deleted
+def clean_up_user_images(self):
```
This is really weird to me. I've never seen self used this way. Why not just use an argument that defaults to false?

```
def clean_up_user_images(user=None):
    img_prefix = get_user_img_prefix(user) if user else temp_img_prefix
```

ajbozarth replied:
after reading through my code a bit I realized that using `self` this way is a remnant of my process to get this feature working, I'll update it so the original handler grabs the cookie out of `self` and passes that instead, in which case I'll be passing the cookie around in the same way. That'll be more proper python code. (I think I originally grabbed more than just the cookie from `self` at one point)

stevemart commented on app.py:
```diff
@@ -174,13 +208,28 @@ def prepare_metadata():
+            image_captions.pop(img_file)
+
+
+# Deletes expired user uploaded files and removes them from the dict
```
same docstring comment as above

stevemart commented on app.py:
```diff
@@ -174,13 +208,28 @@ def prepare_metadata():
+# User uploaded images expire after one day
+def clean_up_old_images():
     img_list = get_image_list()
+    exp_time = time.time() - (24 * 60 * 60)
```
leave a comment here for what `24 * 60 * 60` actually means for future maintainers

ajbozarth replied:
Whoops, must've deleted that, I had it at one point

stevemart commented on app.py:
```diff
@@ -119,13 +133,33 @@ def post(self):
+
+
+def valid_user_img(self, img):
+    default_img = not img.startswith(static_img_path + temp_img_prefix)
```
Can you add a docstring here? There's some funkiness going on that's not easy to follow. Explaining what the method actually does would be helpful.

ajbozarth replied:
will do

stevemart commented on app.py:
```diff
@@ -119,13 +133,33 @@ def post(self):
 
 def valid_file_ext(filename):
-    return '.' in filename and filename.split('.', 1)[1].lower() in VALID_EXT
+    filename, file_extension = os.path.splitext(filename)
```
nit, unused variables part of tuples should be prefixed with an underscore. `_filename, file_extension`

ajbozarth replied:
didn't know that, I'd seen that before but didn't know why

stevemart commented on app.py:
```diff
@@ -119,13 +133,33 @@ def post(self):
 def valid_file_ext(filename):
-    return '.' in filename and filename.split('.', 1)[1].lower() in VALID_EXT
+    filename, file_extension = os.path.splitext(filename)
+    return file_extension.lower() in VALID_EXT
```
add logging here? would have helped with your other bug.

ajbozarth replied:
I had a `print` while testing, but it didn't seem worth adding a `log` line. What string would you recommend to log here? I don't want to cluster the logs too much.

stevemart replied:
log when it's not valid? i'm assuming that would have caught the `.` in filename bug ?

```
if file_extension.lower() not in VALID_EXT
  log something
return
```

ajbozarth replied:
ah, gotcha

stevemart commented on static/js/webapp.js:
```diff
@@ -201,6 +201,19 @@ function select_all(bool) {
 }
 
+function clean_up_imgs() {
+    if (confirm("Are you sure you wish to delete your uploaded images?")) {
```
UX rule 1 (http://babich.biz/ux-writing/) if you can say it in fewer words...

"Delete all uploaded images?"

ajbozarth replied:
I can change it, I was going for the "Are you sure..." from OS empty trash prompts as a kind of joke, but now that this text is on the main page and not the admin page it should probably be clearer.

stevemart commented on templates/index.html:
```diff
@@ -122,7 +129,11 @@ <h1>MAX Image Caption Generator</h1>
+            <p style="float:left">Created at the IBM Center for Open-Source Data & AI Technologies</p>
+            <p style="float:right">
+                Source code available on
+                <a href="https://github.com/IBM/MAX-Image-Caption-Generator-Web-App" target="_blank">GitHub</a>
```
+1! nice addition

ajbozarth replied:
It's in my other web apps so I figured I should add it here too

stevemart commented on tests/test.py:
```diff
@@ -48,7 +53,7 @@ def test_server():
     with open(test_file, 'rb') as file:
         file_form = {'file': file}
-        r = requests.post(url=upload_url, files=file_form)
+        r = requests.post(url=upload_url, files=file_form, cookies=cookie)
```
it seems excessive to add an argument to each test. is it necessary for all of them? they should still pass without one, right?

ajbozarth replied:
I thought that too at first but for a test to see files uploaded in a previous test it needs the same cookie, therefore each test needs to set the cookie. This also allows tests to test user personas this way, there are two "users" in the tests: cookie and cookie2

ajbozarth commented:
@stevemart I've responded to all you comments and will address code updates tomorrow

ajbozarth commented:
@stevemart I ended up just making those changes tonight if you want to take a look.

ajbozarth commented:
Also just realized I need to update the README to remove the Troubleshooting section about the cleanup page and add one about how user sessions work (specifically how they aren't secure). I'll do it in the morning.

frreiss commented on app.py:
```diff
@@ -39,43 +40,60 @@
+    def prepare(self):
+        # Documentation on cookies:
+        # http://www.tornadoweb.org/en/stable/guide/security.html
+        if not self.get_cookie(app_cookie):
```
You probably want to use `get_secure_cookie()` here to prevent advanced users from viewing each others' images. See http://www.tornadoweb.org/en/stable/guide/security.html

ajbozarth replied:
I considered that, but if I use the secure cookie then I can't use it's value in the user's uploaded file names since it's no longer a clean string. In which case I would have to keep a data structure containing a list of users and their files, complicating the code quite a bit.

MLnick replied:
Could you create a simple UUID to store in the secure cookie to use as the filename string? Then it would be a one-to-one from the secure cookie to the uuid

ajbozarth replied:
Based on a better understanding of the library I was able to switch back to secure cookies, just pushed the update for review @MLnick @frreiss I also made the cookie unique to eat instance as well

frreiss commented on app.py:
```diff
@@ -39,43 +40,60 @@
-        self.render("index.html", image_captions=image_captions)
+        clean_up_old_images()
+        cookie = self.get_cookie(app_cookie)
+        self.render("index.html", image_captions=get_image_captions(cookie))
```
You'll want to add a warning about cookies to the bottom of the page for EU users; see https://www.howtogeek.com/327268/why-do-some-websites-have-pop-up-warnings-about-cookies/

You can use the cookie to store whether the user has agreed to your use of cookies.

ajbozarth replied:
Hmmm, I'll look into this, if I prompt the user and they say no then I'd have to disable upload.

frreiss replied:
I'd say that just disabling the entire app if they say no is fine.

ajbozarth replied:
In that case that'll be easy to implement, I can include it in my BaseHandler

ajbozarth commented:
I have a fix for the GDPR compliance mostly ready, but it currently uses the JS `confirm` prompt, so I need to update that to some sort of banner prompt before it's ready. If @MLnick and @frreiss ok my update to use secure cookies before I finish then I will merge this and open up a follow up PR with GDPR compliance as @stevemart recommended

ajbozarth commented:
I just pushed my EU Cookie Consent code, the update includes:

- Adding the 3rd party library Cookie Consent to handle the consent banner
- Switched implementation from using user_id cookie directly to using Tornado's auth user features to access it.
- Added `login` endpoint to initialize a user "session" id, login is triggered on cookie consent.
- Added tests for new endpoint

I've also updated the max-demo instance for testing, note that cookies are unique to an instance of the web app and are changed (but not deleted) on server restart.

ajbozarth commented:
With the two recent updates adding secure cookies and cookie consent functionality could everyone take another look at this? Again the dev endpoint is available for testing: http://max-demo1.fyre.ibm.com:8088 (you may need to clear your resources to refresh the js/css)
@djalova @MLnick @frreiss @bdwyer2 @stevemart @rhagarty 
Thanks for everyone's patience, this feature-add will be big.

frreiss commented:
LGTM
